tests/HttpClient/Plugin/HybridOauth2Test.php

```diff
@@ -0,0 +1,198 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apigee\Edge\Tests\HttpClient\Plugin;
+
+use Apigee\Edge\Client;
+use Apigee\Edge\ClientInterface;
+use Apigee\Edge\HttpClient\Utility\Builder;
+use Apigee\Edge\Tests\Test\HttpClient\MockHttpClient;
+use Apigee\Edge\Tests\Test\HttpClient\Plugin\InMemoryOauthTokenStorage;
+use Apigee\Edge\Tests\Test\HttpClient\Plugin\MockHybridOauth2;
+use Apigee\Edge\Tests\Test\HttpClient\Utility\TestJournal;
+use GuzzleHttp\Psr7\Response;
+use Http\Client\Exception\TransferException;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class HybridOauth2Test.
+ *
+ * @group client
+ * @small
+ */
+class HybridOauth2Test extends TestCase
+{
+    private const API_ENDPOINT = 'http://api.example.com/v1';
+
+    /** @var \Apigee\Edge\Tests\Test\HttpClient\MockHttpClient */
+    protected static $httpClient;
+
+    /** @var \Apigee\Edge\HttpClient\Utility\JournalInterface */
+    private $journal;
+
+    /** @var \Apigee\Edge\ClientInterface */
+    private $client;
+
+    /** @var \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface */
+    private $token_storage;
+
+    /**
+     * @inheritdoc
+     */
+    public static function setUpBeforeClass(): void
+    {
+        // Use the Mock HTTP Client for all requests.
+        self::$httpClient = new MockHttpClient();
+        parent::setUpBeforeClass();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->token_storage = new InMemoryOauthTokenStorage();
+        $this->journal = new TestJournal();
+        $this->client = $this->buildClient();
+    }
+
+    /**
+     * @expectedException \Apigee\Edge\Exception\HybridOauth2AuthenticationException
+     * @expectedException \Apigee\Edge\Exception\OauthAccessTokenAuthenticationException
+     */
+    public function testAuthServerError(): void
+    {
+        // Auth server is unavailable.
+        static::$httpClient->addException(new TransferException());
+        $this->client->get('');
+    }
+
+    public function testOauthReAuthenticationAfterExpiredAccessToken(): void
+    {
+        $body = [
+            'access_token' => 'access_token',
+            'expires_in' => 60,
+            'token_type' => 'Bearer',
+        ];
+        // Auth server returns a new access token.
+        static::$httpClient->addResponse(new Response(200, ['Content-Type' => 'application/json'], json_encode((object) $body)));
+
+        // Successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
+
+        // Another successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
```

Comment:
Why do you have to test the same thing twice?

Reply:
See https://github.com/apigee/apigee-client-php/pull/71#discussion_r341357331